pkg/sharky/shard.go

```diff
@@ -5,187 +5,140 @@
+			// this condition checks if an slot is in limbo (popped but not used for write op)
+			if writes != nil {
+				sh.slots.wg.Add(1) // Done after the slots process pops from slots.in
+				go func() {
+					sh.slots.in <- slot
+				}()
 			}
 			return
-		default:
-			select {
-
-			// pop a free offset
-			case offset = <-free:
-				// only if there is one can we pop a chunk to write otherwise keep back pressure on writes
-				// effectively enforcing another shard to be chosen
-				writeOps = sh.writeOps // enable popping a write operation
-				free = nil             // disabling getting a new position until a write is actually done
-
-				// only enabled if there is a free offset (pos) previously popped
-			case op := <-writeOps:
-				sh.write(op, offset)
-				free = sh.free // reenable popping a free slot next time we can write
-				writeOps = nil // disable popping a write operation until there is a free slot
-
-				// prioritise read operations
-			case op := <-sh.readOps:
-				sh.read(op)
-
-			case <-sh.quit:
-				// again put back offset in limbo
-				if writeOps != nil {
-					sh.freed <- offset
-				}
-				return
-			}
 		}
-
 	}
 }
 
+// close closes the shard:
+// wait for pending operations to finish then saves free slots and blobs on disk
 func (sh *shard) close() error {
-	go func() { // close free channel only after all relevant inflight change
-		sh.wg.Wait()
-		close(sh.freed)
-	}()
-	free := []int64{}
-	for offset := range sh.freed {
-		free = append(free, offset)
-	}
-	frees, err := json.Marshal(free)
-	if err != nil {
+	sh.slots.wg.Wait()
+	if err := sh.slots.save(); err != nil {
 		return err
 	}
-	if err = sh.ffh.Truncate(0); err != nil {
+	if err := sh.slots.file.Close(); err != nil {
 		return err
 	}
-	if _, err = sh.ffh.Seek(0, 0); err != nil {
-		return err
-	}
-	if _, err := sh.ffh.Write(frees); err != nil {
-		return err
-	}
-	if err := sh.ffh.Close(); err != nil {
-		return err
-	}
-	return sh.fh.Close()
+	return sh.file.Close()
 }
 
-// not  called concurrently
-func (sh *shard) read(op *operation) {
-	_, err := sh.fh.ReadAt(op.buffer[:op.location.Length], op.location.Offset)
-	select {
-	case op.err <- err:
-	case <-sh.quit:
-	}
+// offset calculates the offset from the slot
+// this is possible since all blobs are of fixed size
+func (sh *shard) offset(slot uint32) int64 {
+	return int64(slot) * int64(sh.datasize)
 }
 
-// not  called concurrently
-func (sh *shard) write(op *operation, offset int64) {
-	sh.wg.Add(1)
-	_, err := sh.fh.WriteAt(op.data, offset)
-	if err != nil {
-		go func() {
-			sh.freed <- offset
-			sh.wg.Done()
-		}()
-	} else {
-		defer sh.wg.Done()
-		op.location.Offset = offset
-		op.location.Shard = sh.index
-		op.location.Length = int64(len(op.data))
-	}
-	select {
-	case op.err <- err:
-	case <-sh.quit:
+// read reads loc.Length bytes to the buffer from the blob slot loc.Slot
+func (sh *shard) read(r read) error {
+	_, err := sh.file.ReadAt(r.buf, sh.offset(r.slot)+int64(r.at))
```

Comment:
offset(at) part can be omitted as it is not used and is fairly easy to add later on